### PR TITLE
SWATCH-967: Populate subscription records in swatch-contracts

### PR DESCRIPTION
**Pull request by kahowell**
Jira issue: [SWATCH-967](https://issues.redhat.com/browse/SWATCH-967)

Description
===========

Every time a contract is created or updated, a corresponding subscription record is created or updated. Note that because we use `subscriptionId` as part of the primary key, I had to add a client for subscription service. When the subscription record does not exist yet, it is pulled in via subscription number. If the lookup fails, then neither contract nor subscription will be persisted.

There were also a couple of key attributes needed from tag profile, the dimension to metricId mapping, and the billing factor. For example, with rosa, if a contract comes in for 100 `four_vcpu_hours`, then gets translated into 400 `CORES`

Testing the contracts changes
=============================

Setup
-----

1. Set up an ActiveMQ broker for testing:

```shell
podman run -it --rm -p 8161:8161 -p 61616:61616 -p 5672:5672 -e AMQ_EXTRA_ARGS='--java-options -Dbrokerconfig.maxDiskUsage=100' -e AMQ_USER=quarkus -e AMQ_PASSWORD=quarkus quay.io/artemiscloud/activemq-artemis-broker:0.1.4
```

2. Set up wiremock:

```shell
./gradlew swatch-contracts:wiremock
```

Steps
-----

1. Run the service:

```shell
SWATCH_CONTRACT_PRODUCER_ENABLED=true \
  QUARKUS_OPENTELEMETRY_ENABLED=false \
  UMB_NAMESPACE=localtest \
  UMB_ENABLED=true \
  SWATCH_SELF_PSK=placeholder \
  SWATCH_INTERNAL_SUBSCRIPTION_ENDPOINT=http://localhost:8101/mock/internalSubs \
  MP_MESSAGING_INCOMING_CONTRACTS_CONNECTOR=smallrye-amqp \
  MP_MESSAGING_OUTGOING_CONTRACTSTEST_CONNECTOR=smallrye-amqp \
  ENTITLEMENT_GATEWAY_URL=http://localhost:8101/mock/partnerApi \
  SUBSCRIPTION_URL=http://localhost:8101/mock/subscription \
  ENABLE_SPLUNK_HEC=false \
  ./gradlew swatch-contracts:quarkusDev
```

Verification
------------

1. Query the contracts tables and see that there are records:

```shell
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions <<EOF
select *
from contracts c
         join contract_metrics m on c.uuid = m.contract_uuid;
EOF
```

2. Query the subscriptions table and see that the metric ID has been translated back to actual metric IDs according to the response of the tag metrics endpoint, and that billing factor was applied:

```shell
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions <<EOF
select *
from subscription s
         join subscription_measurements m on s.subscription_id = m.subscription_id and s.start_date = m.start_date
         join subscription_product_ids on s.subscription_id = subscription_product_ids.subscription_id and
                                          s.start_date = subscription_product_ids.start_date;
EOF
```

Testing the tag metrics endpoint
================================

Steps
-----

1. Run the service:

```shell
DEV_MODE=true \
  ./gradlew :bootRun
```

Verification
------------

1. Hit the new http endpoint and see the response.

```shell
http :8000/api/rhsm-subscriptions/v1/internal/tags/rosa/metrics \
  x-rh-swatch-psk:placeholder
```

### CODE REVIEW

**Comment by kahowell** [comment]
Note: blocked by #2224

**Comment by Aurobinda55** [comment]
- Created one contract for rosa (four_vcpu_hour=10, control_plane=10)
- Checked contracts and contracts_metrics and got the correct record
- Checked the above query to get the subscription data and got the correct response and the metric value with billing_factor applied.
- Checked new tag metrics endpoint and got the correct response mentioned above.
- Updated the contract_metrics table with a new four_vcpu_hour value=20

```
swatch-tally-db=# select * from contract_metrics;
            contract_uuid             |   metric_id    | metric_value 
--------------------------------------+----------------+--------------
 8fd01a94-e9f8-4047-9183-c4fdfb6c7060 | control_plane  |           10
 8fd01a94-e9f8-4047-9183-c4fdfb6c7060 | four_vcpu_hour |           20
```

- Checked the above query mentioned for subscription value. It shows the old value and is not updated 
```
- swatch-tally-db=# select *               
from subscription s
         join subscription_measurements m on s.subscription_id = m.subscription_id and s.start_date = m.start_date
         join subscription_product_ids on s.subscription_id = subscription_product_ids.subscription_id and
                                          s.start_date = subscription_product_ids.start_date;
-[ RECORD 1 ]-------+-------------------------------------
sku                 | 9bbdb3d1-f574-43de-a298-78623f83c6ee
org_id              | 3340851
subscription_id     | 8fd01a94-e9f8-4047-9183-c4fdfb6c7060
quantity            | 1
start_date          | 2023-06-13 14:16:38+00
end_date            | 
billing_provider_id | 
account_number      | 
subscription_number | cff74b3f-24c6-4817-819c-ea9be673f946
billing_provider    | aws
billing_account_id  | test-125
has_unlimited_usage | f
subscription_id     | 8fd01a94-e9f8-4047-9183-c4fdfb6c7060
start_date          | 2023-06-13 14:16:38+00
metric_id           | redhat.com:rosa:cluster_hour
measurement_type    | PHYSICAL
value               | 10
product_id          | rosa
subscription_id     | 8fd01a94-e9f8-4047-9183-c4fdfb6c7060
start_date          | 2023-06-13 14:16:38+00
-[ RECORD 2 ]-------+-------------------------------------
sku                 | 9bbdb3d1-f574-43de-a298-78623f83c6ee
org_id              | 3340851
subscription_id     | 8fd01a94-e9f8-4047-9183-c4fdfb6c7060
quantity            | 1
start_date          | 2023-06-13 14:16:38+00
end_date            | 
billing_provider_id | 
account_number      | 
subscription_number | cff74b3f-24c6-4817-819c-ea9be673f946
billing_provider    | aws
billing_account_id  | test-125
has_unlimited_usage | f
subscription_id     | 8fd01a94-e9f8-4047-9183-c4fdfb6c7060
start_date          | 2023-06-13 14:16:38+00
metric_id           | redhat.com:rosa:cpu_hour
measurement_type    | PHYSICAL
value               | 40
product_id          | rosa
subscription_id     | 8fd01a94-e9f8-4047-9183-c4fdfb6c7060
start_date          | 2023-06-13 14:16:38+00

swatch-tally-db=#  

```

@kahowell  Could you please confirm if this is these are the correct steps or if I am missing anything?

**Comment by kahowell** [comment]
> Checked the above query mentioned for subscription value. It shows the old value and is not updated.

@Aurobinda55 this is probably an issue we should address. Do you happen to have a reproducer? Otherwise can you please elaborate on how you updated the contract_metrics table?

**Comment by Aurobinda55** [comment]
> This is probably an issue we should address. Do you happen to have a reproducer? Otherwise can you please elaborate on how you updated the contract_metrics table?

Reproducer 
- Create a contract for rosa (four_vcpu_hour=10, control_plane=10)
- Query the contracts table and subscription table and check the data
- From the swatch-tally-db pod, update the contract_metrics table with updated metrics value (four_vcpu_hour=20, control_plane=20)
- Check contract_metrics and it shows the updated value
- Check the subscription table value for the metric_value and it shows the old values i.e 10,40(applied billing factor)

**Comment by kahowell** [comment]
> From the [swatch-tally-db](https://issues.redhat.com/browse/SWATCH-tally-db) pod, update the contract_metrics table with updated metrics value (four_vcpu_hour=20, control_plane=20)

@Aurobinda55 , I don't expect a direct database update to trigger recalculation of the subscription measurements values. I would expect a contract update via messaging or API to update them though.

**Comment by kahowell** [comment]
79821a7 to 95a96d9: no-change rebase to get fresh CI runs

**Comment by Aurobinda55** [comment]
- Checked the new metrics endpoint and got the below response for rosa 
`[{"uom":"INSTANCE_HOURS","aws_dimension":"control_plane","billing_factor":1.0},{"uom":"CORES","aws_dimension":"four_vcpu_hour","billing_factor":0.25}]`

- Added a contract (four_vcpu_hour=10,control_plane=10)
- Checked the contract metrics and got the expected value 
- Checked the subscription table and got the actual metric id and it matched with the response from the new metrics endpoint  i.e metric_id=CORES, value =40, metric_id=INSTANCE_HOURS=10
- Removed the contract to see the changes in the contract and subscription table 
- it removed from the subscription and contract table

**Comment by Aurobinda55** [comment]
> > From the [swatch-tally-db](https://issues.redhat.com/browse/SWATCH-tally-db) pod, update the contract_metrics table with updated metrics value (four_vcpu_hour=20, control_plane=20)
> 
> @Aurobinda55 , I don't expect a direct database update to trigger recalculation of the subscription measurements values. I would expect a contract update via messaging or API to update them though.

@kahowell  the update contract api is already removed, so I was not able to find a way to update the contract through the API.  other scenarios are looks good.

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/model/MeasurementMetricIdTransformer.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import javax.inject.Inject;
+import lombok.extern.slf4j.Slf4j;
+import org.eclipse.microprofile.rest.client.inject.RestClient;
+
+@Slf4j
+@ApplicationScoped
+public class MeasurementMetricIdTransformer {
+  @RestClient @Inject InternalSubscriptionsApi internalSubscriptionsApi;
+
+  public void translateContractMetricIdsToSubscriptionMetricIds(SubscriptionEntity subscription) {
```
Comment (lindseyburnett) [inline swatch-contracts/src/main/java/com/redhat/swatch/contract/model/MeasurementMetricIdTransformer.java:41]: Javadoc for this

**Review by lindseyburnett — APPROVED** [review]

**Pull request merged**